source/core_sntp_client.c

```diff
@@ -102,3 +121,293 @@ SntpStatus_t Sntp_Init( SntpContext_t * pContext,
+    assert( pContext->authIntf.generateClientAuth != NULL );
+
+    status = pContext->authIntf.generateClientAuth( pContext->authIntf.pAuthContext,
+                                                    &pContext->pTimeServers[ pContext->currentServerIndex ],
+                                                    pContext->pNetworkBuffer,
+                                                    pContext->bufferSize,
+                                                    &authDataSize );
+
+    if( status != SntpSuccess )
+    {
+        LogError( ( "Unable to send time request: Client authentication function failed: "
+                    "RetStatus=%u", status ) );
+    }
+
+    /* Sanity check that the returned authentication data size fits in the remaining space
+     * of the request buffer besides the first #SNTP_PACKET_BASE_SIZE bytes. */
+    else if( authDataSize > ( pContext->bufferSize - SNTP_PACKET_BASE_SIZE ) )
+    {
+        LogError( ( "Unable to send time request: Invalid authentication code size: "
+                    "AuthCodeSize=%lu, NetworkBufferSize=%lu",
+                    ( unsigned long ) authDataSize, ( unsigned long ) pContext->bufferSize ) );
+        status = SntpErrorAuthFailure;
+    }
+    else
+    {
+        /* With the authentication data added. calculate total SNTP request packet size. The same
+         * size would be expected in the SNTP response from server. */
+        pContext->sntpPacketSize = SNTP_PACKET_BASE_SIZE + authDataSize;
+
+        LogInfo( ( "Appended client authentication code to SNTP request packet:"
+                   " AuthCodeSize=%lu, TotalPacketSize=%lu",
+                   ( unsigned long ) pContext->sntpPacketSize,
+                   ( unsigned long ) pContext->sntpPacketSize ) );
+    }
+
+    return status;
+}
+
+SntpStatus_t Sntp_SendTimeRequest( SntpContext_t * pContext,
+                                   uint32_t randomNumber )
+{
+    SntpStatus_t status = SntpSuccess;
+
+    if( pContext == NULL )
+    {
+        status = SntpErrorBadParameter;
+        LogError( ( "Invalid context parameter: Context cannot be NULL" ) );
+    }
+    else
+    {
+        const SntpServerInfo_t * pServer = NULL;
+
+        /* Check if there is any time server available for requesting time
+         * that has not already rejected a prior request. */
+        if( pContext->currentServerIndex >= pContext->numOfServers )
+        {
+            LogError( ( "Cannot request time: All servers have rejected time requests: "
+                        "Re-initialize context with new servers" ) );
+            status = SntpErrorChangeServer;
+        }
+        else
+        {
+            /* Set local variable for the currently indexed server to use for time
+             * query. */
+            pServer = &pContext->pTimeServers[ pContext->currentServerIndex ];
+
+            LogDebug( ( "Using server %.*s for time query", ( int ) pServer->serverNameLen, pServer->pServerName ) );
+        }
```

Comment:
Nit: You can decrease nesting here by moving the index check a level above
```suggestion
    /* Check if there is any time server available for requesting time
     * that has not already rejected a prior request. */
    else if( pContext->currentServerIndex >= pContext->numOfServers )
    {
        LogError( ( "Cannot request time: All servers have rejected time requests: "
                    "Re-initialize context with new servers" ) );
        status = SntpErrorChangeServer;
    }
    else
    {
        const SntpServerInfo_t * pServer = NULL;

        /* Set local variable for the currently indexed server to use for time
         * query. */
        pServer = &pContext->pTimeServers[ pContext->currentServerIndex ];

        LogDebug( ( "Using server %.*s for time query", ( int ) pServer->serverNameLen, pServer->pServerName ) );
```